### PR TITLE
store and retrieve offchain proofs

### PR DESCRIPTION
**Pull request by fforbeck**
**Saves the merkle root and steps of the merkle tree into the Snapshot-Hub DB if the steps and merkle root are valid**
 
`POST /:space/offchain_proofs` - d 

- Request
```json
  "verifyingContract": "...",
  "actionId": "...",
  "chainId": "...",
  "merkleRoot": "...",
  "steps": [
   {
    "nbYes": "...", 
    "nbNo": "...",
    "account": "...",
    "choice": "...",
    "proposalHash": "...",
    "index": "..."
    }
  ]
```

**Retrieves the steps of the merkle tree based on the space and merkle root**
`GET /:space/offchain_proof/:merkleRoot`

* Response
```json
  "merkleRoot": "...",
  "steps": [
   {
    "nbYes": "...", 
    "nbNo": "...",
    "account": "...",
    "choice": "...",
    "proposalHash": "...",
    "index": "..."
    }
  ]
```

### CODE REVIEW

**Comment by jtrein** [comment]
Thank you, @fforbeck. The issue referenced above is the one I will work on to use this PR. Will keep this PR updated with any progress.

**Comment by jtrein** [comment]
Do I need to run any DB migration? When running this branch with a fresh Docker container I get the following error:

```
snapshot-postgres    | 2021-05-04 14:23:29.615 GMT [140] ERROR:  syntax error at or near "(" at character 1315
snapshot-postgres    | 2021-05-04 14:23:29.615 GMT [140] STATEMENT:  CREATE TABLE IF NOT EXISTS messages (
snapshot-postgres    | 	  id VARCHAR(66) NOT NULL,
snapshot-postgres    | 	  address VARCHAR(64) NOT NULL,
snapshot-postgres    | 	  version VARCHAR(6) NOT NULL,
snapshot-postgres    | 	  timestamp BIGINT NOT NULL,
snapshot-postgres    | 	  space VARCHAR(64),
snapshot-postgres    | 	  token VARCHAR(64),
snapshot-postgres    | 	  type VARCHAR(12) NOT NULL,
snapshot-postgres    | 	  payload JSON,
snapshot-postgres    | 	  sig VARCHAR(256) NOT NULL,
snapshot-postgres    | 	  metadata JSON,
snapshot-postgres    | 	  "actionId" VARCHAR(64) NOT NULL,
snapshot-postgres    | 	  data JSONB,
snapshot-postgres    | 	  PRIMARY KEY (id)
snapshot-postgres    | 	);
snapshot-postgres    |
snapshot-postgres    | 	CREATE INDEX IF NOT EXISTS mgs_address_idx on messages (address);
snapshot-postgres    | 	CREATE INDEX IF NOT EXISTS mgs_version_idx on messages (version);
snapshot-postgres    | 	CREATE INDEX IF NOT EXISTS mgs_timestamp_idx on messages (timestamp);
snapshot-postgres    | 	CREATE INDEX IF NOT EXISTS mgs_space_idx on messages (space);
snapshot-postgres    | 	CREATE INDEX IF NOT EXISTS mgs_token_idx on messages (token);
snapshot-postgres    | 	CREATE INDEX IF NOT EXISTS mgs_type_idx on messages (type);
snapshot-postgres    | 	CREATE INDEX IF NOT EXISTS mgs_action_idx on messages ("actionId");
snapshot-postgres    |
snapshot-postgres    | 	CREATE TABLE IF NOT EXISTS hubs (
snapshot-postgres    | 	  host VARCHAR(64) NOT NULL,
snapshot-postgres    | 	  address VARCHAR(64),
snapshot-postgres    | 	  is_self INT DEFAULT 0,
snapshot-postgres    | 	  is_active INT DEFAULT 1,
snapshot-postgres    | 	  PRIMARY KEY (host)
snapshot-postgres    | 	);
snapshot-postgres    |
snapshot-postgres    | 	CREATE INDEX IF NOT EXISTS hubs_address_idx on hubs (address);
snapshot-postgres    | 	CREATE INDEX IF NOT EXISTS hubs_is_self_idx on hubs (is_self);
snapshot-postgres    |
snapshot-postgres    | 	CREATE TABLE IF NOT EXISTS offchain_proofs (
snapshot-postgres    | 	  space VARCHAR(64) NOT NULL,
snapshot-postgres    | 	  merkle_root VARCHAR(64) NOT NULL,
snapshot-postgres    | 	  steps JSONB NOT NULL,
snapshot-postgres    | 	  created_at DATETIME NOT NULL DEFAULT CURRENT_DATETIME()
snapshot-postgres    | 	  PRIMARY KEY (space, merkle_root)
snapshot-postgres    | 	);
snapshot-postgres    |
snapshot-postgres    | 	CREATE INDEX IF NOT EXISTS offchain_proofs_mklr_idx on offchain_proofs (merkle_root);
snapshot-postgres    | 	CREATE INDEX IF NOT EXISTS offchain_proofs_space_idx on offchain_proofs (space);
snapshot-hub         |
snapshot-hub         | /home/node/app/node_modules/pg-protocol/src/parser.ts:357
snapshot-hub         |         : new DatabaseError(messageValue, length, name)
snapshot-hub         |           ^
snapshot-hub         | error: syntax error at or near "("
snapshot-hub         |     at Parser.parseErrorMessage (/home/node/app/node_modules/pg-protocol/src/parser.ts:357:11)
snapshot-hub         |     at Parser.handlePacket (/home/node/app/node_modules/pg-protocol/src/parser.ts:186:21)
snapshot-hub         |     at Parser.parse (/home/node/app/node_modules/pg-protocol/src/parser.ts:101:30)
snapshot-hub         |     at Socket.<anonymous> (/home/node/app/node_modules/pg-protocol/src/index.ts:7:48)
snapshot-hub         |     at Socket.emit (events.js:315:20)
snapshot-hub         |     at Socket.EventEmitter.emit (domain.js:485:12)
snapshot-hub         |     at addChunk (_stream_readable.js:296:12)
snapshot-hub         |     at readableAddChunk (_stream_readable.js:272:9)
snapshot-hub         |     at Socket.Readable.push (_stream_readable.js:213:10)
snapshot-hub         |     at TCP.onStreamRead (internal/stream_base_commons.js:186:23)
```

**Comment by fforbeck** [comment]
@jtrein just fixed that.

File: server/index.ts [hunk 1]
```diff
@@ -189,6 +193,69 @@ router.get('/:space/proposal/:id/votes', async (req, res) => {
...
+      buildVoteLeafHashForMerkleTree(s, verifyingContract, actionId, chainId)
+    )
+  );
+
+  if (merkleTree.getHexRoot() === merkleRoot) {
+    await saveOffchainProof(space, merkleRoot, steps);
+    return res.status(201).send();
+  }
+
+  return res.status(403).send({ error: 'invalid merkle root' });
```
Comment (jtrein) [inline server/index.ts:246]: Should this also be a `BadRequest` `400` status code like the others?

File: server/index.ts [hunk 2]
```diff
@@ -189,6 +193,69 @@ router.get('/:space/proposal/:id/votes', async (req, res) => {
...
+});
+
+router.get('/:space/offchain_proof/:merkleRoot', async (req, res) => {
+  const { space, merkleRoot } = req.params;
+  return getOffchainProof(space, merkleRoot)
+    .then(p => res.status(200).send(p))
+    .catch(e => {
+      console.error(e);
+      return res.status(404).send();
```
Comment (jtrein) [inline server/index.ts:255]: Perhaps this should be a `500`, and inside the `.then(p => ...)` if the the `p.length === 0` (I think it's an `Array`?) then it should `res.status(404).send()` then `return`?

File: server/index.ts [hunk 3]
```diff
@@ -189,6 +193,69 @@ router.get('/:space/proposal/:id/votes', async (req, res) => {
...
+  const steps = req.body.steps;
+
+  if (
+    !space ||
+    !verifyingContract ||
+    !actionId ||
+    !chainId ||
+    !merkleRoot ||
+    !steps ||
+    steps.length < 1 // must have at least 2 steps
```
Comment (jtrein) [inline server/index.ts:211]: Is the comment correct here? Maybe `// must have at least 1 step` ?

File: server/index.ts [hunk 4]
```diff
@@ -189,6 +193,69 @@ router.get('/:space/proposal/:id/votes', async (req, res) => {
...
+  )
+    return res.status(400).send({
+      error: 'invalid request parameters'
+    });
+
+  const indexes = steps.map(s => s.index).filter(s => s.index);
```
Comment (jtrein) [inline server/index.ts:217]: Are `s.index` a `number`? If so, `0` will not pass the `filter`. E.g.:

```js
[{index: 0}].map(s => s.index).filter(s => s.index)

[]
```

**Review by jtrein — CHANGES_REQUESTED** [review]
Hey, @fforbeck I've seen a couple small things for your review.

File: server/index.ts (hunk 1 above)
Comment (fforbeck) [reply server/index.ts]: Fixed.

File: server/index.ts (hunk 2 above)
Comment (fforbeck) [reply server/index.ts]: Yeah, it is an array. Just changed it accordingly.

File: server/index.ts (hunk 3 above)
Comment (fforbeck) [reply server/index.ts]: ah, it should be < 2, because with only 1 step it is not valid. Fixed.

File: server/index.ts (hunk 4 above)
Comment (fforbeck) [reply server/index.ts]: great catch. Thanks. Just fixed that.

**Comment by fforbeck** [comment]
@jtrein thanks for reviewing the PR. I just fixed the issues, it should be ready for another review round.

File: server/index.ts [hunk 5]
```diff
@@ -189,6 +193,72 @@ router.get('/:space/proposal/:id/votes', async (req, res) => {
...
+  }
+
+  return res.status(400).send({ error: 'invalid merkle root' });
+});
+
+router.get('/:space/offchain_proof/:merkleRoot', async (req, res) => {
+  const { space, merkleRoot } = req.params;
+  return getOffchainProof(space, merkleRoot)
+    .then(p => {
+      if (p && p.length > 0) return res.status(200).send(p);
```
Comment (jtrein) [inline server/index.ts:253]: Sorry, is there any way you can unpack the array to return the single, unique result? And if `404` then just `res.sendStatus(400)`?

**Review by jtrein — COMMENTED** [review]
Could we make this change to the `GET`?

File: server/index.ts (hunk 5 above)
Comment (fforbeck) [reply server/index.ts:253]: absolutely

Comment (fforbeck) [reply server/index.ts]: fixed.

**Comment by jtrein** [comment]
@fforbeck I changed some HTTP response status things, FYI. Let me know if that's OK.

**Comment by fforbeck** [comment]
> @fforbeck I [changed](https://github.com/openlawteam/snapshot-hub/pull/35/commits/f82fb771b21a6f4d23aa124eed507f2aeaf7ccb8) some HTTP response status things, FYI. Let me know if that's OK.

it looks great @jtrein

**Comment by jtrein** [comment]
@fforbeck Could you review, please? I made some more changes.

**Review by jtrein — APPROVED** [review]
Looks good!

Manually tested:

- `POST` with 1 step
- `POST` with steps > 1
- `GET`
- `GET` 404

**Pull request merged**